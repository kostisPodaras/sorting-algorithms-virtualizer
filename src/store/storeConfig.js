import { routerReducer } from 'core/components/router';

import { combineEpics, epics as coreEpics } from 'core/models/epics';

import { epics as userEpics, reducer as userReducer } from 'models/user';

const initialReducer = {
  router: routerReducer,
  user: userReducer,
};

const initialEpics = combineEpics(coreEpics, userEpics);

export { initialEpics, initialReducer };
