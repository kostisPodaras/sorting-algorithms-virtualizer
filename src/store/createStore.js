import { applyMiddleware, createStore as reduxCreateStore } from 'redux';
import { createEpicMiddleware } from 'redux-observable';
import { composeWithDevTools } from 'redux-devtools-extension';

import { createHistory, routerMiddleware } from 'core/components/router';
import { compose, storageValue } from 'core/utils';
import { combineReducers, REDUCER } from 'core/models/reducer';

import {
  attachEpicMiddlewareToStore,
  initializeEpics,
} from 'core/models/epics';

import config from 'config';

import { initialEpics, initialReducer } from './storeConfig';

const { get: getInitialState, set: setStateInStorage } = storageValue(
  config.siteShortName,
  true,
);

let storeCache;

const createStore = (preloadedState) => {
  if (storeCache) {
    return storeCache;
  }

  const epicMiddleware = createEpicMiddleware();
  const middleware = [epicMiddleware, routerMiddleware];

  const composeEnhancers = process.env.DEPLOY_BUILD
    ? compose
    : composeWithDevTools({
        name: 'Agile Actors',
        trace: true,
        traceLimit: 25,
      });

  const store = reduxCreateStore(
    combineReducers(initialReducer),
    getInitialState() || preloadedState || {},
    composeEnhancers(applyMiddleware(...middleware)),
  );

  attachEpicMiddlewareToStore(store, epicMiddleware);
  initializeEpics(store, initialEpics);

  store[REDUCER] = initialReducer;
  store.history = createHistory(store);

  let previousState = preloadedState || {};

  store.subscribe(() => {
    const newState = store.getState();

    previousState !== newState && setStateInStorage(newState);
    previousState = newState;
  });

  storeCache = store;

  return store;
};

export default createStore;
