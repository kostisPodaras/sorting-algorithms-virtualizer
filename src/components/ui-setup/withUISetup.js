import React from 'react';

import { createStore } from 'store';

const withUISetup = (Component) => (props) => {
  const store = createStore();
  const { history } = store;

  return <Component history={history} {...props} />;
};

export default withUISetup;
