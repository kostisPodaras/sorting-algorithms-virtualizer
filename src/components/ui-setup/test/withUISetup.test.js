import * as store from 'store';

import withUISetup from '../withUISetup';

describe('withUISetup HOC', () => {
  test('renders successfully', () => {
    // GIVEN
    const createStoreSpy = jest.spyOn(store, 'createStore').mockReturnValue({
      history: [],
    });
    const DummyComponent = () => <div />;
    const WrappedDummyComponent = withUISetup(DummyComponent);

    // WHEN
    const tree = shallowSnapshot(<WrappedDummyComponent />);

    // THEN
    expect(createStoreSpy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });
});
