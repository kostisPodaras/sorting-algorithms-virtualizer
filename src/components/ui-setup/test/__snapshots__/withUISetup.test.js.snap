// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUISetup HOC renders successfully 1`] = `
<DummyComponent
  history={Array []}
/>
`;
